src/asset/asset-list-loader.js

```diff
@@ -0,0 +1,215 @@
+     * @param {Asset[]|number[]} assetList - An array of {@link Asset} objects to load or an array of Asset IDs to load.
+     * @param {AssetRegistry} assetRegistry - The application's asset registry.
+     * @example
+     * const assetListLoader = new pc.AssetListLoader([
+     *     new pc.Asset("texture1", "texture", { url: 'http://example.com/my/assets/here/texture1.png') }),
+     *     new pc.Asset("texture2", "texture", { url: 'http://example.com/my/assets/here/texture2.png') })
+     * ], pc.app.assets);
+     */
+    constructor(assetList, assetRegistry) {
+        super();
+        this._assets = [];
+        this._loadingAssets = new Set();
+        this._registry = assetRegistry;
+        this._loaded = false;
+        this._count = 0; // running count of successfully loaded assets
+        this._failed = []; // list of assets that failed to load
+
+        this._waitingAssets = [];
+
+        if (assetList.length && assetList[0] instanceof Asset) {
+            assetList.forEach((asset, i, array) => {
+                // filter out duplicates
+                if (array.indexOf(asset) !== i) {
+                    return;
+                }
+                if (!asset.registry) {
+                    asset.registry = assetRegistry;
+                }
+                this._assets.push(asset);
+            });
+        } else {
+            // list of Asset IDs
+            assetList.forEach((assetId, i, array) => {
+                // filter out duplicates
+                if (array.indexOf(assetId) !== i) {
+                    return;
+                }
+                const asset = assetRegistry.get(assetList[i]);
+                if (asset) {
+                    this._assets.push(asset);
+                } else {
+                    this._waitForAsset(assetList[i]);
+                }
+            });
+        }
+    }
+
+    /**
+     * Removes all references to this asset list loader.
+     */
+    destroy() {
+        // remove any outstanding listeners
+        const self = this;
+
+        this._registry.off("load", this._onLoad);
+        this._registry.off("error", this._onError);
+
+        this._waitingAssets.forEach(function (id) {
+            self._registry.off("add:" + id, this._onAddAsset);
+        });
+
+        this.off("progress");
+        this.off("load");
+    }
+
+    /**
+     * Start loading asset list, call done() when all assets have loaded or failed to load.
+     *
+     * @param {Function} done - Callback called when all assets in the list are loaded. Passed (err, failed) where err is the undefined if no errors are encountered and failed contains a list of assets that failed to load.
+     * @param {object} [scope] - Scope to use when calling callback.
+     */
+    load(done, scope) {
+        const l = this._assets.length;
+        let asset;
+
+        this._count = 0;
+        this._failed = [];
+        this._callback = done;
+        this._scope = scope;
+        this._loaded = false;
```

Comment:
this block suggests the function can be called multiple times, but I'm not sure it should, it that it would work?

Reply:
I've made this function return with a debug message if it's called multiple times. I've also added in a test case for multiple function calls.